File: C# Samples/Advanced/Datasets/UsingDatasets/RunUsingDatasets.cs


using System;
using System.IO;
using System.Collections.Generic;
using WindwardReportsDrivers.net.windward.datasource;
using WindwardReportsDrivers.net.windward.datasource.ado;
using WindwardReportsDrivers.net.windward.datasource.xml;
using WindwardInterfaces.net.windward.api.csharp;
using net.windward.api.csharp;

namespace UsingDatasets
{
	/// <summary>
	/// Sample code to set datasets (SQL & XML) using a .rdlx file.
	/// </summary>
	public class RunUsingDatasets
	{
		/// <summary>
		/// Sample code to set datasets (SQL & XML) using a .rdlx file.
		/// </summary>
		/// <param name="args">nothing</param>
		static void Main(string[] args){
		
			// Initialize the engine
            Report.Init();

            // Open template file and create output file
			using (FileStream template = new FileStream("../../files/Sample Dataset Template.docx", 
													FileMode.Open, FileAccess.Read, FileShare.Read))
			{
				using (FileStream output = new FileStream("../../files/Sample Dataset Report.pdf",
													FileMode.Create, FileAccess.Write, FileShare.None))
				{

					// Create report process
					using (Report myReport = new ReportPdf(template, output))
					{
						// read in the template
						myReport.ProcessSetup();

						// XML datasource
						using (FileStream xmlFile = new FileStream("../../files/SouthWind.xml", FileMode.Open, FileAccess.Read, FileShare.Read))
						using (FileStream xmlSchema = new FileStream("../../files/SouthWind.xsd", FileMode.Open, FileAccess.Read, FileShare.Read))
						using (SaxonDataSourceImpl dsSaxon = new SaxonDataSourceImpl(xmlFile, xmlSchema))
						using (DataSetImpl dsEmployeesUnder5 = new DataSetImpl("employeesUnder5", "/windward-studios/Employees/Employee[@EmployeeID < 5]", dsSaxon))
            			using (DataSetImpl dsCustStartA = new DataSetImpl("CustStartA", "/windward-studios/Customers/Customer[starts-with(CompanyName, 'A')]", dsSaxon))
						// SQL datasource
						using (AdoDataSourceImpl dsAdo = new AdoDataSourceImpl("System.Data.SqlClient", "Data Source=mssql.windward.net;Initial Catalog=Northwind;User ID=demo;Password=demo"))
			            using (DataSetImpl dsEmployeesMoreThan5 = new DataSetImpl("EmpMoreThan5", "SELECT * FROM dbo.Employees WHERE(dbo.Employees.EmployeeID > 5)", dsAdo))
            			using (DataSetImpl dsCustStartWithB = new DataSetImpl("CustStartWithB", "SELECT * FROM dbo.Customers WHERE(dbo.Customers.CompanyName like 'B%')", dsAdo))
						{
							IDictionary<string, IReportDataSource> datasources = new Dictionary<string, IReportDataSource>();
							datasources.Add("SW", dsSaxon);
            				datasources.Add("employeesUnder5", dsEmployeesUnder5);
            				datasources.Add("CustStartA", dsCustStartA);
            				datasources.Add("MSSQL", dsAdo);
				            datasources.Add("EmpMoreThan5", dsEmployeesMoreThan5);
				            datasources.Add("CustStartWithB", dsCustStartWithB);

							myReport.ProcessData(datasources);
						}

						// all data applied, finish up the report.
						myReport.ProcessComplete();

						// no need to call close because of the using constructs
					}
				}
			}

			// Opens the finished report
			string fullPath = Path.GetFullPath("../../files/Sample Dataset Report.pdf");
			Console.Out.WriteLine(string.Format("launching {0}", fullPath));
            System.Diagnostics.Process.Start(fullPath);
		}
	}
}

File: Java Samples/Advanced/Datasets/UsingDatasets/src/RunUsingDatasets.java
/*
* Copyright (c) 2012 by Windward Studios, Inc. All rights reserved.
*
* This program can be copied or used in any manner desired.
*/

import net.windward.datasource.*;
import net.windward.datasource.dataset.DataSetDataSource;
import net.windward.datasource.xml.*;
import net.windward.datasource.jdbc.*;
import net.windward.env.SystemWrapper;
import net.windward.env.WindwardWrapper;
import net.windward.xmlreport.ProcessPdf;
import net.windward.xmlreport.ProcessReportAPI;
import net.windward.env.SystemWrapper;

import java.io.*;
import java.lang.reflect.Method;
import java.util.HashMap;
import java.util.Map;

public class RunUsingDatasets {

    public static void main(String[] args) throws Exception {

        try {
            try {
                Class.forName("com.microsoft.sqlserver.jdbc.SQLServerDriver").newInstance();
            } catch (ClassNotFoundException e) {
                throw new ClassNotFoundException("Please add the SqlServer JDBC connector to your classpath. Details at http://rpt.me/SqlServerJDBC", e);
            }

            // To generate a report, first we need a ProcessReport object.
            // For now, we're using the pdf format to output.
            InputStream streamTemplate = new FileInputStream("files/Sample Dataset Template.docx");
            File reportFile = new File("out/Sample Dataset Report.pdf");
            OutputStream streamReport = new FileOutputStream(reportFile);


            ProcessReportAPI report = new ProcessPdf(streamTemplate, streamReport);

            // Preparation...
            System.out.println("Generating report...");
            report.processSetup();
            streamTemplate.close();

            // apply the XML data
            InputStream streamXmlData = new FileInputStream("files/SouthWind.xml");
            InputStream streamXmlSchema = new FileInputStream("files/SouthWind.xsd");
            DataSourceProvider dsSaxon = new SaxonDataSource(streamXmlData, streamXmlSchema);
            DataSetDataSource dsEmployeesUnder5 = new DataSetDataSource("employeesUnder5", "/windward-studios/Employees/Employee[@EmployeeID < 5]", dsSaxon);
            DataSetDataSource dsCustStartA = new DataSetDataSource("CustStartA", "/windward-studios/Customers/Customer[starts-with(CompanyName, 'A')]", dsSaxon);

            // apply the SQL data
            DataSourceProvider dsJdbc = new JdbcDataSource("com.microsoft.sqlserver.jdbc.SQLServerDriver",
                    "jdbc:sqlserver://mssql.windward.net;DatabaseName=Northwind", "demo", "demo");
            DataSetDataSource dsEmployeesMoreThan5 = new DataSetDataSource("EmpMoreThan5", "SELECT * FROM dbo.Employees WHERE(dbo.Employees.EmployeeID > 5)", dsJdbc);
            DataSetDataSource dsCustStartWithB = new DataSetDataSource("CustStartWithB", "SELECT * FROM dbo.Customers WHERE(dbo.Customers.CompanyName like 'B%')", dsJdbc);

            Map<String, DataSourceProvider> datasources = new HashMap<String, DataSourceProvider>();
            datasources.put("SW", dsSaxon);
            datasources.put("employeesUnder5", dsEmployeesUnder5);
            datasources.put("CustStartA", dsCustStartA);
            datasources.put("MSSQL", dsJdbc);
            datasources.put("EmpMoreThan5", dsEmployeesMoreThan5);
            datasources.put("CustStartWithB", dsCustStartWithB);
            report.processData(datasources);

            streamXmlData.close();
            streamXmlSchema.close();

            // And... DONE!
            report.processComplete();
            streamReport.close();

            // launch the generated report
            System.out.println("all threads completed");
            System.out.println("Launching report " + reportFile.getAbsolutePath());
            SystemWrapper.LaunchFile(reportFile.getAbsolutePath());

        } catch (Exception e) {
            // Uh oh, just in case
            e.printStackTrace();
        }
    }
}
